File: sdk/typescript/src/contract/clustering-id-migration.contract.test.ts
import * as fs from 'fs';
import * as path from 'path';
import { fallbackHeal } from '../core-fallback';
import type { DOMSnapshot, HealResult, Signature } from '../core-types';

const FIXTURE_DIR = path.resolve(
  __dirname,
  '../../../../fixtures/contract/clustering-id-migration'
);

interface ContractExpected {
  healed: boolean;
  min_confidence: number;
  best_candidate_attribute: string;
  best_candidate_value: string;
}

interface HealRequestFixture {
  primary_selector: string;
  original_signature: Signature;
  dom_snapshot: DOMSnapshot;
  min_confidence: number;
}

function loadFixture<T>(name: string): T {
  return JSON.parse(fs.readFileSync(path.join(FIXTURE_DIR, name), 'utf-8')) as T;
}

function assertContract(result: HealResult, expected: ContractExpected): void {
  expect(result.healed).toBe(expected.healed);
  expect(result.confidence).toBeGreaterThanOrEqual(expected.min_confidence);
  expect(result.top_candidates.length).toBeGreaterThan(0);

  const best = result.top_candidates[0];
  const needle = `${expected.best_candidate_attribute}="${expected.best_candidate_value}"`;
  expect(best.selector).toContain(needle);

  if (result.healed) {
    expect(result.selector).toContain(needle);
  }
}

describe('contract: clustering id → data-id migration', () => {
  it('fallbackHeal matches shared fixture expectations', () => {
    const request = loadFixture<HealRequestFixture>('request.json');
    const expected = loadFixture<ContractExpected>('expected.json');

    const result = fallbackHeal(
      request.primary_selector,
      request.original_signature,
      request.dom_snapshot,
      request.min_confidence
    );

    assertContract(result, expected);
  });
});

File: sdk/java/frap-core-java/src/test/java/io/github/kotlerdev/frap/core/client/ClusteringIdMigrationContractTest.java
package io.github.kotlerdev.frap.core.client;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.github.kotlerdev.frap.core.dto.Candidate;
import io.github.kotlerdev.frap.core.dto.HealRequest;
import io.github.kotlerdev.frap.core.dto.HealResult;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.file.Path;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Cross-language contract: id → data-id migration in the same list (fixtures/contract).
 */
class ClusteringIdMigrationContractTest {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private FrapCoreClient client;

    @BeforeEach
    void setUp() throws IOException {
        client = new FrapRpcClient(FrapCoreBinaryResolver.resolve());
    }

  @org.junit.jupiter.api.AfterEach
  void tearDown() {
    if (client != null) {
      client.close();
    }
  }

    @Test
    void healsWhenLiIdMigratesToDataId() throws IOException {
        Path fixtureDir = Path.of("../../../fixtures/contract/clustering-id-migration");
        HealRequest request = MAPPER.readValue(fixtureDir.resolve("request.json").toFile(), HealRequest.class);
        JsonNode expected = MAPPER.readTree(fixtureDir.resolve("expected.json").toFile());

        HealResult result = client.heal(request);

        assertThat(result.healed()).isEqualTo(expected.get("healed").asBoolean());
        assertThat(result.confidence()).isGreaterThanOrEqualTo(expected.get("min_confidence").asDouble());
        assertThat(result.topCandidates()).isNotEmpty();

        Candidate best = result.topCandidates().get(0);
        String attr = expected.get("best_candidate_attribute").asText();
        String value = expected.get("best_candidate_value").asText();
        assertThat(best.selector()).contains(attr + "=\"" + value + "\"");

        if (result.healed()) {
            assertThat(result.selector()).contains(attr + "=\"" + value + "\"");
        }
    }
}
